### ISSUE TITLE
Bug: [strict-void-return] false positive when function has overload with void-returning function and Promise-returning function

### ISSUE THREAD

**Issue by reduckted** [typescript-eslint/typescript-eslint#11960]
### Before You File a Bug Report Please Confirm You Have Done The Following...

- [x] I have tried restarting my IDE and the issue persists.
- [x] I have updated to the latest version of the packages.
- [x] I have [searched for related issues](https://github.com/typescript-eslint/typescript-eslint/issues?q=is%3Aissue+label%3A%22package%3A+eslint-plugin%22) and found none that matched my issue.
- [x] I have [read the FAQ](https://typescript-eslint.io/linting/troubleshooting) and my problem is not listed.

### Playground Link

https://typescript-eslint.io/play/#ts=5.5.2&fileType=.tsx&code=JYOwLgpgTgZghgYwgAgBIHt0GsBiBXEBMYdEAHgBVkIAPSEAEwGdkA3dYB5AHzUy2QBeNhwYA%2BZAG8AsAChkyABQwQALiUBKIRPacN6igG45C5Ws3bkABSjoAtsCYQyu8fuRG5AXzlyEAGzgmFgxsKR9ZOQYIALgoFFY45AAjCBh0eIBRRAALdVDcAiIScgKxY0jZVPSs3MVFLUEJSWQvDQq5aoyIbIQcxSCAT0ILJqlW9rkgA&eslintrc=N4KABGBEBOCuA2BTAzpAXGUEKQAIBcBPABxQGNoBLY-AWhXkoDt8B6ZfKsugNwHtKAE1rRE%2BWNCbooAdwCGkyODABfECqA&tsconfig=N4KABGBEDGD2C2AHAlgGwKYCcDyiAuysAdgM6QBcYoEEkJemy0eAcgK6qoDCAFutAGsylBm3TgwAXxCSgA&tokens=false

### Repro Code

```TypeScript
interface HookFunction<T extends void | Hook = void> {
  (fn: () => void): T;
  (fn: () => Promise<void>): T;
}

class Hook {}

declare var beforeEach: HookFunction<Hook>;

beforeEach(() => {});

beforeEach(async () => { });

```

### ESLint Config

```javascript
module.exports = {
  parser: "@typescript-eslint/parser",
  rules: {
    "@typescript-eslint/strict-void-return": "warn"
};
```

### tsconfig

```jsonc
{
  "compilerOptions": {
    "strictNullChecks": true
  }
}
```

### Expected Result

`beforeEach(async () => { });` should be valid because the parameter type is a function that returns a `Promise`.

### Actual Result

`beforeEach(async () => { });` produces a warning because the parameter is considered to be a function that returns void.

### Additional Info

The specific case where I encountered this problem is with Mocha's `beforeEach` method.

https://github.com/DefinitelyTyped/DefinitelyTyped/blob/49a2604fdf497676f6bc005b4869729d5b0d05bd/types/mocha/index.d.ts#L392

**Comment by reduckted** [comment]
https://github.com/typescript-eslint/typescript-eslint/blob/3021ede961732e66b1daca804759aa0cf4143b97/packages/eslint-plugin/src/rules/strict-void-return.ts#L133

🤔 When `tsNode` is `async () => {}`, `expectedType` is `() => void`.

**Comment by JoshuaKGoldberg** [comment]
Fun! It's nifty how commenting out the `(fn: () => void): T;` makes the `beforeEach(async () => { });` pass. 👍 from me on this being a bug to be fixed. Thanks for filing!
